### DISCUSSION TITLE
Testing service with multilevel services dependencies

### DISCUSSION THREAD

**Discussion by akbarsahata** [L2jLiga/fastify-decorators#659] [Q&A] [answered]
Hi, I am writing a test for service with dependency of other service with dependency of some service too.

Below are the dependencies

```typescript
@Service()
export class MySQLConnection {
  connection!: Connection;

  @Initializer()
  async init() {
    this.connection = await createConnection({
      type: 'mysql',
      database: mysqlEnv.MYSQL_DB,
      host: mysqlEnv.MYSQL_HOST,
      port: mysqlEnv.MYSQL_PORT,
      username: mysqlEnv.MYSQL_USERNAME,
      password: mysqlEnv.MYSQL_PASSWORD,
      synchronize: false,
      ....
```
```typescript
@Service()
export class VariableDAO {
  @Inject(MySQLConnection)
  private mySql!: MySQLConnection;

  private repository!: Repository<Variable>;

  @Initializer([MySQLConnection])
  async init() {
    this.repository = this.mySql.connection.getRepository(Variable);
  }

  async getAll(params?: FindManyOptions<Variable>): Promise<Variable[]> {
    return this.repository.find(params);
  }
}
```
And below is the service I'd like to write a test

```typescript
@Service()
export class VariableService {
  @Inject(VariableDAO)
  private variableDAO!: VariableDAO;

  async getAll(): Promise<Variable[]> {
    return this.variableDAO.getAll();
  }
}
```

And here's the test
```typescript
describe('Service: VariableService', () => {
  let service: VariableService;
  const variableDAO = { getAll: jest.fn() };

  beforeEach(async () => {
    service = configureServiceTest({
      service: VariableService,
      mocks: [
        {
          provide: VariableDAO,
          useValue: variableDAO,
        },
      ],
    });
  });

  afterEach(() => jest.restoreAllMocks());

  it('should return array of Variable', async () => {
    variableDAO.getAll.mockReturnValue([]);

    const result = await service.getAll();

    expect(result).toHaveLength(0);
  });
});
```
When I ran the test, it seems like the real instance of `VariableDAO` is called rather than the mocked one. The error is from MySQLConnection class where I did some process.env validation.

I already configured Jest as told in documentation. Did I miss something?

I know it's unrelated, but here is the error returned by Jest. The point is, it seemed like the real instance dependency of VariableDAO is called. anyway, apologize for my bad english

```
  ●  process.exit called with "1"

       5 | import { cleanEnv, str, port, bool } from 'envalid';
       6 | 
    >  7 | export const mysqlEnv = cleanEnv(process.env, {
         |                                 ^
       8 |   MYSQL_DB: str(),
       9 |   MYSQL_HOST: str(),
      10 |   MYSQL_PORT: port({ default: 3306 }),

      at defaultReporter (node_modules/envalid/src/reporter.ts:67:13)
      at Object.getSanitizedEnv (node_modules/envalid/src/core.ts:100:3)
      at cleanEnv (node_modules/envalid/src/envalid.ts:18:19)
      at Object.<anonymous> (src/config/datasource.ts:7:33)
      at Object.<anonymous> (src/datasources/connection/mysql.connection.ts:4:1)
```

**Comment by unematiii** [comment]
Could it be that beforeEach has an async callback, but you are really not await-ing for anything?

**Comment by L2jLiga** [comment, marked as answer]
It's seems like you've missed environment configuration, it's caused because `VariableService` imports `VariableDAO` which imports `MySQLCommection` service which imports your `cleanEnv` and even `VariableDAO` is mocked and not actually instantiated - imports are still loaded and hence your `cleanEnv` executed which leads to this validation error, nothing wrong with library or mocks, you have to configure your environment variables before tests, e.g. in `jest.config.js`

See https://dev.to/darkmavis1980/how-to-mock-environmental-variables-with-jest-2j3c for example

**Reply by L2jLiga** [reply]
Another option is to create service which will validate environment variables on instantiation, e.g.:

```typescript
@Service()
export class EnvironmentService {
  // Don't forget your type here
  get environment () {
    return this._environment;
  }
  private _environment;

  constructor () {
    this._environment = cleanEnv()
  }

}
```


This way you'll still have working cleanEnv with validation and option to mock environment by mocking this service or any services depending on this one

**Reply by akbarsahata** [reply]
thanks for pointing that the import is still loaded. setting up env vars in `jest.config.js` is way easier. really appreciate your answer :pray:
